src/rewards/FlywheelGaugeRewards.sol

```diff
@@ -0,0 +1,173 @@
+    If `e` is the cycle end, `t` is the min of e and current timestamp, and `p` is the prior updated time:
+    For `A` accrued rewards over the cycle, distribute `min(A * (t-p)/(e-p), A)`.
+*/ 
+contract FlywheelGaugeRewards is Auth, IFlywheelRewards {
+    using SafeTransferLib for ERC20;
+
+    // TODO add events
+
+    /// @notice the flywheel core contract
+    address public immutable flywheel;
+
+    /// @notice the reward token paid
+    ERC20 public immutable rewardToken;
+
+    /// @notice the end of the current cycle
+    uint32 public rewardsCycleEnd;
+
+    /// @notice the length of a rewards cycle
+    uint32 public immutable rewardsCycleLength;
+
+    /// @notice rewards queued from prior and current cycles
+    struct QueuedRewards {
+        uint128 priorCycleRewards;
+        uint128 currentCycleRewards;
+    }
+
+    /// @notice mapping from gauges to queued rewards
+    mapping(ERC20 => QueuedRewards) public marketQueuedRewards;
+
+    /// @notice the gauge token for determining gauge allocations of the rewards stream
+    ERC20Gauges public immutable gaugeToken;
+
+    /// @notice contract to pull reward tokens from
+    IRewardsStream public rewardsStream;
+
+    constructor(
+        ERC20 _rewardToken, 
+        address _flywheel, 
+        address _owner, 
+        Authority _authority,
+        uint32 _rewardsCycleLength,
+        ERC20Gauges _gaugeToken,
+        IRewardsStream _rewardsStream
+    ) Auth(_owner, _authority) {
+        rewardToken = _rewardToken;
+        flywheel = _flywheel;
+        rewardsCycleLength = _rewardsCycleLength;
+
+        // seed initial rewardsCycleEnd
+        rewardsCycleEnd = uint32(block.timestamp) / rewardsCycleLength * rewardsCycleLength;
+
+        gaugeToken = _gaugeToken;
+
+        rewardsStream = _rewardsStream;
+    }
+
+    /**
+        @notice Iterates over all live gauges and queues up the rewards for the cycle
+        @return totalQueuedForCycle the max amount of rewards to be distributed over the cycle
+        @dev critical that there is no pagination to prevent double spending. All gauges must be synced atomically, unless the gaugeToken checkpoints rewards.
+
+        GAS: Average path for `n` gauges - SLOAD * (n + 7) + SSTORE * (n + 1) + 1 token transfer
+    */
+    function queueRewardsForCycle() external requiresAuth returns (uint256 totalQueuedForCycle) {
+        // ensure new cycle has begun
+        require(block.timestamp > rewardsCycleEnd); // SLOAD
+        
+        // next cycle is always the next even divisor of the cycle length above current block timestamp.
+        rewardsCycleEnd = (uint32(block.timestamp) + rewardsCycleLength) / rewardsCycleLength * rewardsCycleLength; // SSTORE
+
+        // queue the rewards stream and sanity check the tokens were received
+        uint256 balanceBefore = rewardToken.balanceOf(address(this)); // SLOAD
+        totalQueuedForCycle = rewardsStream.getRewards();
```

Comment:
Minor point, but on first pass I read this `getRewards()` as a pure getter and was confused as to why the token balance of this contract would increase. 

Perhaps just me but I associate functions pre-fixed with `get` as pure getters/view functions. Maybe `transferRewards()` or similar?

Reply:
how about queueRewards?